src/libraries/System.Linq/src/System/Linq/Sum.cs

```diff
@@ -49,6 +72,103 @@ public static partial class Enumerable
+            Debug.Assert(Vector<T>.Count > 2);
+            Debug.Assert(Vector.IsHardwareAccelerated);
+
+            ref T ptr = ref MemoryMarshal.GetReference(span);
+            int length = span.Length;
+
+            // Overflow testing for vectors is based on setting the sign bit of the overflowTracking
+            // vector for an element if the following are all true:
+            //   - The two elements being summed have the same sign bit. If one element is positive
+            //     and the other is negative then an overflow is not possible.
+            //   - The sign bit of the sum is not the same as the sign bit of the previous accumulator.
+            //     This indicates that the new sum wrapped around to the opposite sign.
+            //
+            // By bitwise or-ing the overflowTracking vector for each step we can save cycles by testing
+            // the sign bits less often. If any iteration has the sign bit set in any element it indicates
+            // there was an overflow.
+            //
+            // Note: The overflow checking in this algorithm is only correct for signed integers.
+            // If support is ever added for unsigned integers then the overflow check should be:
+            //   overflowTracking |= (accumulator & data) | Vector.AndNot(accumulator | data, sum);
+
+            Vector<T> accumulator = Vector<T>.Zero;
+
+            // Build a test vector with only the sign bit set in each element. JIT will fold this into a constant.
+            Vector<T> overflowTestVector = new(T.RotateRight(T.MultiplicativeIdentity, 1));
+
+            // Unroll the loop to sum 4 vectors per iteration
+            do
+            {
+                // Switch accumulators with each step to avoid an additional move operation
+                Vector<T> data = Vector.LoadUnsafe(ref ptr);
+                Vector<T> accumulator2 = accumulator + data;
+                Vector<T> overflowTracking = (accumulator2 ^ accumulator) & (accumulator2 ^ data);
+
+                data = Vector.LoadUnsafe(ref ptr, (nuint)Vector<T>.Count);
+                accumulator = accumulator2 + data;
+                overflowTracking |= (accumulator ^ accumulator2) & (accumulator ^ data);
+
+                data = Vector.LoadUnsafe(ref ptr, (nuint)Vector<T>.Count * 2);
+                accumulator2 = accumulator + data;
+                overflowTracking |= (accumulator2 ^ accumulator) & (accumulator2 ^ data);
+
+                data = Vector.LoadUnsafe(ref ptr, (nuint)Vector<T>.Count * 3);
+                accumulator = accumulator2 + data;
+                overflowTracking |= (accumulator ^ accumulator2) & (accumulator ^ data);
+
+                if ((overflowTracking & overflowTestVector) != Vector<T>.Zero)
+                {
+                    ThrowHelper.ThrowOverflowException();
+                }
+
+                ptr = ref Unsafe.Add(ref ptr, Vector<T>.Count * 4);
+                length -= Vector<T>.Count * 4;
```

Comment:
Similar feedback has been given on other PRs, but we'd prefer to avoid mutating the `byref` where possible in favor of standard indexing and using `LoadUnsafe(ref baseAddress, index)` instead. You can then increment `index` by `Vector<T>.Count` in the loop.

This can result in slightly worse codegen in some cases, but it makes the code simpler and less prone to GC holes and similar bugs.

Reply:
Done, and the JIT output still looks decent. It was able to encode the offset arithmetic into the operation on x64:

```asm
lea ebx, [eax+edi+0x40]
```

Unfortunately, it's an extra `lea` operation. While this did impact the benchmarks, they are still an improvement over unvectorized.